### ISSUE TITLE
Change Password Hash to SHA2

### ISSUE THREAD

**Issue by IronManMark20** [r0m30/msed#10]
As Bruce Schneier showed ( and many others have suggested) [SHA-1 will be broken in the next 3-6 years](https://www.schneier.com/blog/archives/2012/10/when_will_we_se.html). I am hoping that in a new release for future proofing, you could update for SHA-2 so that this awesome program doesn't go the way of the Dodo.

**Comment by r0m30** [comment]
Yes, stronger hashes would be harder to break and if I could find documented test cases and Open source code I could update the PBKDF2 to SHA2 or whatever the currently agreed upon safe hash is. 
I strongly believe in using existing (and hopefully vetted) code for these types of tasks.  

Something to remember, msed doesn't store the key, it passes it to the firmware, and the firmware stores it in one of it's non-host accessible tables.  Since we are looking at black box encryption we have no idea what, if anything, the firmware does to the password after it is passed to it, if it does nothing then it really doesn't make any difference what we do to the password as the attacker has the end product of our hashing.  If it does hash it then the attacker has to break that hash as well. 

Finally, if the attacker has breached the firmware we are probably already compromised.

**Comment by IronManMark20** [comment]
Very true, but it seems that taking over the firmware (even considering the most recent revelations), is not that common (and hard to do). Anyway, that isn't a problem that can be dealt with, so we should probably deal with the ones we can.

 I think the safest and easiest to implement would be the PBKDF2+SHA-512. Considering [this](https://stackoverflow.com/questions/26311686/bcrypt-vs-pbkdf2withhmacsha1), SHA-512's 64bit nature makes it less likely to be weaker to GPU attacks, as GPUs are 32 bit. 

Bcrypt ([see here](https://security.stackexchange.com/questions/4781/do-any-security-experts-recommend-bcrypt-for-password-storage)) seems to be strong and resistant to GPU attacks, but you would have to re-implement msed, and it has shown to be weaker to FPGAs.

Finally, as for being vetted, SHA-512 has been in use since 2002, and bcrypt since 1999. That is long enough for me.

**Comment by r0m30** [comment]
I agree and I thought I saw somewhere in the spec that OPAL drive firmware
needs to be signed.  I wonder if JTAG is also disabled.

As I said before I have no objection to strengthening the hash. I could use
this <http:///> to validate a SHA512 implementation. I'll put it on the
list.

I was referring to the code implementing the algorithm being vetted not
implying the algorithms you were suggesting where not vetted.

----------
Mike

**Comment by IronManMark20** [comment]
Of course! Sorry. I might recommend [crypto++](http://www.cryptopp.com/). It is fairly mature. They have all SHA-2 implementations, and the license implies that the source files are public domain.

BTW your link was broken, could you repost it? Thanks! I think I will install msed then.

**Comment by r0m30** [comment]
I started with crypto++ specifically because it was PD.  When I got to the
PBA it had to be in C so I changed to the GNULIB code.

Fixed the link, apparently github borked the link, probably a problem
parsing the email.

----------
Mike

**Comment by CheckQ** [comment]
Hello, I've found information regarding "SHA-1 is broken" in the cryptsetup (dm-crypt/LUKS) FAQ 5.20 LUKS is broken! It uses SHA-1!

A small piece of the answer:
SHA-1 is (academically) broken for finding collisions, but not for using it in a key-derivation function.

For the full answer read:
https://gitlab.com/cryptsetup/cryptsetup/wikis/FrequentlyAskedQuestions

You are using SHA-1 in a key-derivation function. So everything seems to be fine.

**Comment by r0m30** [comment]
Thanks for that info,  I think I will probably still leave this open as an enhancement request.

**Comment by praetoralpha** [comment]
Do not do this! SHA2 is not designed to store password hashes. PBKDF2 is.

**Comment by r0m30** [comment]
I don't think anyone was suggesting not to use PBKDF2, just to use a stronger hash in the PBKDF2 implementation.
